### ISSUE TITLE
`(a%b) == a` Parser error.

### ISSUE THREAD

**Issue by necto** [verifast/verifast#16]
The VeriFast parser gets quire annoying sometimes. E.g.:

This code verifies:

```
/*@
lemma void to_prove(int a, int b);
requires 0 <= a &*& a < b;
ensures a == (a%b);
@*/
```

This does not (note I swapped `a` and `(a%b)` in the last equation):

```
/*@
lemma void to_prove(int a, int b);
requires 0 <= a &*& a < b;
ensures (a%b) == a;
@*/
```

> Parser error.

**Comment by btj** [comment]
Yes. The current parser has the unfortunate "feature" that, when parsing an assertion, if it encounters a left parenthesis, it expects a parenthesized assertion, not just a parenthesized expression.

It is a relatively high-priority TODO item to fix this, probably by merging the "assertion" grammatical category into the "expression" category. That will also get rid of the need to write `f(...) == true` instead of just `f(...)` when using a boolean fixpoint application as an assertion. It will then be the type-checker's job, rather than the parser's, to decide whether `f(...)` is a predicate application or a fixpoint application.

**Comment by btj** [comment]
Fixed by 65df8947773e63e1254c1f7d8093f6801d891b07 .
